File: codes/julia/ch21/tricks_common.jl
# tricks_common.jl
#
# Shared utilities for Chapter 21 (Special Tricks for the Spectral Researcher).
#
# Provides:
#   * NAVY, CORAL, TEAL, PURPLE, GOLD, OLIVE colour constants;
#   * `apply_theme!()` to set CairoMakie theme to CMU Serif at the
#     standard textbook size;
#   * `output_dir(script_dir)` to resolve textbook/figures/ch21/julia/;
#   * `write_json(path, x)` for the etude --dump option.
#
# Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
# Part of "Computational Etudes: A Spectral Approach"

using CairoMakie
using Colors

const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#8E44AD"
const ORANGE = colorant"#E67E22"
const GOLD   = colorant"#D4A017"
const OLIVE  = colorant"#6B8E23"

function apply_theme!()
    set_theme!(Theme(fontsize = 11,
        fonts = (regular = "CMU Serif", bold = "CMU Serif Bold",
                 italic = "CMU Serif Italic"),
        Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 11,
                xticklabelsize = 10, yticklabelsize = 10,
                spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
        Legend = (labelsize = 9,)))
end

function output_dir(script_dir::AbstractString)
    p = script_dir
    while !isempty(p) && !isdir(joinpath(p, "textbook"))
        new_p = dirname(p)
        new_p == p && break
        p = new_p
    end
    out = joinpath(p, "textbook", "figures", "ch21", "julia")
    mkpath(out)
    return out
end

# ---- Minimal JSON writer (no external dependency) ----
_json_val(x::Bool) = x ? "true" : "false"
_json_val(x::Integer) = string(x)
_json_val(x::Real) = isfinite(x) ? string(Float64(x)) :
    (isnan(x) ? "NaN" : (x > 0 ? "Infinity" : "-Infinity"))
_json_val(x::Complex) = "{\"re\": " * _json_val(real(x)) *
                        ", \"im\": " * _json_val(imag(x)) * "}"
_json_val(x::AbstractString) = "\"" *
    replace(String(x), "\\" => "\\\\", "\"" => "\\\"") * "\""
_json_val(v::AbstractVector) = "[" * join(_json_val.(v), ", ") * "]"
_json_val(t::Tuple) = "[" * join(_json_val.(collect(t)), ", ") * "]"
_json_val(d::AbstractDict) = "{" *
    join([_json_val(string(k)) * ": " * _json_val(v) for (k, v) in d], ", ") *
    "}"

write_json(path, x) = open(io -> print(io, _json_val(x)), path, "w")

File: codes/python/ch21/tricks_common.py
#!/usr/bin/env python3
"""
tricks_common.py

Shared utilities for Chapter 21 (Special Tricks for the Spectral Researcher).

Provides:
  * matplotlib setup and the consistent NAVY/CORAL/TEAL/PURPLE palette;
  * output directory resolution to textbook/figures/ch21/python/;
  * tiny helpers (save_fig, dump_json) used by every etude in this chapter.

Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
Part of "Computational Etudes: A Spectral Approach"
"""
from __future__ import annotations

import json
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

NAVY, SKY, CORAL = "#142D6E", "#7896D2", "#E74C3C"
TEAL, PURPLE, ORANGE = "#16A085", "#8E44AD", "#E67E22"
GOLD, OLIVE = "#D4A017", "#6B8E23"

RC = {
    "font.family": "serif",
    "font.serif": ["CMU Serif", "DejaVu Serif"],
    "mathtext.fontset": "cm",
    "font.size": 11,
    "axes.linewidth": 0.8,
    "xtick.major.width": 0.8,
    "ytick.major.width": 0.8,
    "figure.dpi": 150,
    "savefig.dpi": 300,
}


def setup_matplotlib():
    plt.rcParams.update(RC)


def output_dir_for(script_path: Path) -> Path:
    p = script_path.resolve()
    if p.is_file():
        p = p.parent
    while p != p.parent and not (p / "textbook").is_dir():
        p = p.parent
    out = p / "textbook" / "figures" / "ch21" / "python"
    out.mkdir(parents=True, exist_ok=True)
    return out


def save_fig(fig, out_dir: Path, stem: str) -> None:
    for ext in ("pdf", "png"):
        fig.savefig(out_dir / f"{stem}.{ext}", bbox_inches="tight")


def dump_json(path: str | None, payload: dict) -> None:
    if path is None:
        return
    with open(path, "w") as f:
        json.dump(payload, f, indent=2, default=_json_default)


def _json_default(obj):
    if isinstance(obj, (np.floating, np.integer)):
        return float(obj)
    if isinstance(obj, np.ndarray):
        return obj.tolist()
    if isinstance(obj, complex):
        return {"re": obj.real, "im": obj.imag}
    raise TypeError(f"Cannot JSON-serialise {type(obj).__name__}")
